composer/loggers/slack_logger.py

```diff
@@ -0,0 +1,143 @@
+                            'text': f'*{k}:* {v}'
+                        }
+                    } for k, v in data.items()]),
+                    include_keys={'loss/train/total'},
+                    interval_in_seconds=1
+                ),
+            ],
+        )
+
+        trainer.fit()
+
+    Args:
+        formatter_func ((...) -> Any | None): A formatter function that returns list of blocks to be sent to slack.
+        include_keys (Set[str]): A set of metric/logs/traces keys to include in the message. If None, all keys are included.
+        interval_in_seconds (int): Interval in seconds to send logs to slack. If None, message is sent after every log call.
+        max_logs_per_message (int)(default:50): Maximum number of logs to send in a single message. Note that no more than 50 items are allowed to send in a single message.
+        If more than 50 items are stored in buffer, the message flushed without waiting the full time interval.
+    """
+
+    def __init__(
+        self,
+        formatter_func: Optional[Callable[..., List[Dict[str, Any]]]] = None,
+        include_keys: Optional[Set[str]] = None,
+        interval_in_seconds: Optional[int] = None,
+        max_logs_per_message: int = 50,
+    ) -> None:
+        try:
+            import slack_sdk
+            self.client = slack_sdk.WebClient()
+            del slack_sdk
+        except ImportError as e:
+            raise MissingConditionalImportError('slack_logger', 'slack_sdk', None) from e
+
+        self.slack_logging_api_key = os.environ.get('SLACK_LOGGING_API_KEY', None)
+        self.channel_id = os.environ.get('SLACK_LOGGING_CHANNEL_ID', None)
+
+        if self.slack_logging_api_key is None:
+            raise RuntimeError('SLACK_LOGGING_API_KEY must be set as environment variable')
+        if self.channel_id is None:
+            raise RuntimeError('SLACK_LOGGING_CHANNEL_ID must be set as environment variable')
+
+        self.formatter_func = formatter_func
+        self.include_keys = include_keys
+        self.interval_in_seconds = interval_in_seconds
+
+        self.logs, self.last_log_time = [], time.time()
+        self.max_logs_per_message = min(max_logs_per_message, 50)
+
+    # Rich message layouts can be created using Slack Blocks Kit.
+    # See documentation here: https://api.slack.com/messaging/composing/layouts
+    def _log_to_slack(
+            self,
+            data: Dict[str, Any],
+            **kwargs,  # can be used to pass additional arguments to the formatter function (eg for headers)
+    ):
+        filtered_data = {k: v for k, v in data.items() if k in self.include_keys
+                        } if self.include_keys is not None else data
+        blocks = self.formatter_func(
+            filtered_data, **
+            kwargs) if self.formatter_func is not None else self._default_log_bold_key_normal_value_pair_with_header(
+                filtered_data, **kwargs)
+
+        self.logs += blocks
```

Comment:
hmmm I'm not sure appending to a list is the best approach here because if the user sets an infrequent interval, they will see pretty stale data and if they set a high max_logs_per_message they will get overwhelmed with a full history for every key. 
I think we may want to either have a queue or remove values from the list based on a separate dictionary of the keys that are "pending" to be logged and if a new value for that key comes in before that key, value pair gets posted, then we overwrite it?

Reply:
if they set a high `max_logs_per_message`, we will still flush the logs when it hits 50 as that's the maximum that we can send to slack client. 

I am conflicted about having a full history for every key. Conceptually, if we wanted to track something like loss spikes, would we not need to have a full history of total loss at every step? On the other hand, we have also decided against version history for run metadata (although maybe that's more because of technical limitation vs desired product behavior) - cc @mvpatel2000 for thoughts here too. 

Given slack is high noise channel, maybe we'd optimize for small number of keys but larger number of values logged over time, rather than large number of keys and snapshots being logged?

Reply:
Synced quickly with @eracah , we can think about adding `log_full_history: Optional[bool]=False` for future, just launch this first without version history.